### PR TITLE
Unify builders across signals

### LINKED ISSUE #1527: Instrumentation scope - move to builder pattern?
https://github.com/open-telemetry/opentelemetry-rust/pull/1021 added supported for Attributes to instrumentation scope, which already had version, schema_url.
This started with just name and evolved to add more and more (because spec added them). https://github.com/open-telemetry/opentelemetry-rust/pull/1021#discussion_r1161364638 called out the need for potentially making this to a builder pattern, so as to protect us from breaking change in the future, if there are more stuff added to scope.

```
let tracer = TracerProvider.tracer("tracer_name"); // common case. this is still a good idea to keep to keep things simple for majority.
let tracer = TracerProvider::tracer_builder("tracer_name") // advanced scenario.
    .with_version("1.0.0")
    .with_attributes([kv])
    .build();
```

Opening an issue to explore if moving to builder pattern makes sense here (all 3 signals), and if yes, implement the actual change.
Also, there could be other places where such changes are required to make things cleaner/more idiomatic and also protect from future breaking changes.

### LINKED ISSUE #2164: Metrics API - Move to builder pattern for creating Meters
Related to https://github.com/open-telemetry/opentelemetry-rust/issues/1527, but for Metrics
This is part of Metrics API, and hence must be done by Metrics API RC.

### PR DESCRIPTION
**Pull request by stormshield-fabs**
Fixes #1527 and fixes #2164

## Changes

The consensus reached during the SIG meeting is that we'd like to unify the builder APIs across signals, limit duplication and keep our traits object-safe. 
- `opentelemetry`
    - Removed deprecated method `InstrumentationLibrary::new`
    - Renamed `InstrumentationLibrary` to `InstrumentationScope`
    - Renamed `InstrumentationLibraryBuilder` to `InstrumentationScopeBuilder`
    - Removed deprecated methods `LoggerProvider::versioned_logger` and `TracerProvider::versioned_tracer`
    - Removed methods `LoggerProvider::logger_builder`, `TracerProvider::tracer_builder` and `MeterProvider::versioned_meter`
    - Replaced these methods with `LoggerProvider::logger_with_scope`, `TracerProvider::logger_with_scope`, `MeterProvider::meter_with_scope`
    - Replaced `global::meter_with_version` with `global::meter_with_scope`
    - Added `global::tracer_with_scope`
- `opentelemetry_sdk`:
    - Removed `InstrumentationLibrary` re-export and its `Scope` alias, use `opentelemetry::InstrumentationLibrary` instead.
    - Unified builders across signals
      - Removed deprecated `LoggerProvider::versioned_logger`, `TracerProvider::versioned_tracer`
      - Removed `MeterProvider::versioned_meter`
      - Replaced these methods with `LoggerProvider::logger_with_scope`, `TracerProvider::logger_with_scope`, `MeterProvider::meter_with_scope`

## Migration guide (example for `logger` but the changes are similar across all 3 signals)

```rust
let provider = LoggerProvider::builder().build();

// Before:
let library = std::sync::Arc::new(
    InstrumentationLibrary::builder("test_logger")
        .with_version("v1.2.3")
        .with_schema_url("https://opentelemetry.io/schema/1.0.0")
        .build(),
);
let logger = provider.library_logger(library); // no longer available, use provider.logger_with_scope


// or maybe (was deprecated):
let logger = provider.versioned_logger(
  "test_logger",
  Some("v1.2.3".into()),
  Some("https://opentelemetry.io/schema/1.0.0".into()),
);

// After:
let scope = InstrumentationScope::builder("test_logger")
  .with_version("v1.2.3")
  .with_schema_url("https://opentelemetry.io/schema/1.0.0")
  .build();

let logger = provider.logger_with_scope(scope);
```

## Merge requirement checklist

* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-rust/blob/main/CONTRIBUTING.md) guidelines followed
* [ ] Unit tests added/updated (if applicable)
* [x] Appropriate `CHANGELOG.md` files updated for non-trivial, user-facing changes
* [x] Changes in public API reviewed (if applicable)

### CODE REVIEW

File: opentelemetry-appender-log/src/lib.rs
Comment (cijothomas) [inline opentelemetry-appender-log/src/lib.rs:174]: lets keep the with_version as before.

File: opentelemetry-otlp/examples/basic-otlp-http/src/main.rs
Comment (cijothomas) [inline opentelemetry-otlp/examples/basic-otlp-http/src/main.rs:127]:
```suggestion
            .with_attributes(common_scope_attributes)
```

File: opentelemetry-otlp/examples/basic-otlp/src/main.rs
Comment (cijothomas) [inline opentelemetry-otlp/examples/basic-otlp/src/main.rs:116]:
```suggestion
            .with_attributes(common_scope_attributes)
```

File: opentelemetry-otlp/examples/basic-otlp/src/main.rs
Comment (cijothomas) [inline opentelemetry-otlp/examples/basic-otlp/src/main.rs:119]: wondering why tracing is done slightly differently than metric?

File: opentelemetry-stdout/examples/basic.rs
Comment (cijothomas) [inline opentelemetry-stdout/examples/basic.rs:74]: From a useability standpoint, I think its best to not require Arc here. Cloning InstrumentationLibrary should be cheap anyway.

File: opentelemetry/src/global/metrics.rs
Comment (cijothomas) [inline opentelemetry/src/global/metrics.rs:58]: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#get-a-meter

I think we should rename InstrumentationLibrary to InstrumentationScope everywhere, as Scope is what user provides.

**Review by cijothomas — COMMENTED** [review]
Left few comments but looks good overall.
Please check the comments and see if they make sense.

**Comment by cijothomas** [comment]
> There were some concerns about having `Arc` in the `library_signal` APIs, but someone also mentioned we want to avoid unnecessary copies. We could hide the `Arc` inside the `InstrumentationLibrary` if needed.

yes I suggest to hide Arc inside.

File: opentelemetry/src/global/metrics.rs
Comment (cijothomas) [reply opentelemetry/src/global/metrics.rs:58]: @lalitb @utpilla Please share your thoughts on this.

Comment (utpilla) [reply opentelemetry/src/global/metrics.rs:58]: Yes. `InstrumentationScope` also matches the term used in [proto file](https://github.com/open-telemetry/opentelemetry-proto/blob/7312bdf63218acf27fe96430b7231de37fd091f2/opentelemetry/proto/common/v1/common.proto#L69-L81).

Comment (lalitb) [reply opentelemetry/src/global/metrics.rs:58]: +1 for `InstrumentationScope`

Comment (stormshield-fabs) [reply opentelemetry/src/global/metrics.rs]: done

File: opentelemetry-otlp/examples/basic-otlp/src/main.rs
Comment (stormshield-fabs) [reply opentelemetry-otlp/examples/basic-otlp/src/main.rs]: implemented `global::library_tracer`

File: opentelemetry-stdout/examples/basic.rs
Comment (stormshield-fabs) [reply opentelemetry-stdout/examples/basic.rs]: `InstrumentationLibrary` was already cheap to clone, except for the `attributes`. I've put them in a `Cow` so we no longer need to use `Arc`.

File: opentelemetry/src/common.rs
Comment (cijothomas) [inline opentelemetry/src/common.rs:431]: there was a recent clarification that scope attributes also contribute to uniqueness test.
https://github.com/open-telemetry/opentelemetry-specification/blob/main/CHANGELOG.md#logs-1

Let fix in a follow up PR to keep this PR's scope.

File: opentelemetry-stdout/examples/basic.rs
Comment (utpilla) [reply opentelemetry-stdout/examples/basic.rs]: I noticed that you changed it to `pub attributes: Cow<'static, Vec<KeyValue>>`. How does wrapping the vec with `Cow` help?

Did you mean to change that to cow of static slice instead? `Cow<'static, [KeyValue]>`?

Comment (cijothomas) [reply opentelemetry-stdout/examples/basic.rs]: Its okay to not optimize for cloning cost. 
Cloning same instrumentation-scope is to be done at max for 2 times (one for each signal), so it is okay!
(Or on the future, we can do Scope backed by a Arc<InnerScope> so clone is super cheap (if we ever need that))

File: opentelemetry/src/metrics/meter.rs
Comment (utpilla) [inline opentelemetry/src/metrics/meter.rs:28]: Looks like a copy-paste error. Replace logger with `meter` in the comments and variable names.

File: opentelemetry-sdk/src/metrics/meter_provider.rs
Comment (utpilla) [inline opentelemetry-sdk/src/metrics/meter_provider.rs:155]: No need to clone here.

```suggestion
                .entry(scope)
```

File: opentelemetry/src/common.rs
Comment (lalitb) [inline opentelemetry/src/common.rs:431]:
```suggestion
// Uniqueness for InstrumentationScope does not depend on attributes
```

File: opentelemetry/src/common.rs
Comment (cijothomas) [reply opentelemetry/src/common.rs:413]: this should be a relative easy renaming for users, so I don't think its worth maintaining the alias for few releases. 
This is part of `opentelemetry` and we are aiming to declare it RC in a week and stable soon after that.

File: opentelemetry/src/common.rs
Comment (lalitb) [inline opentelemetry/src/common.rs:538]: Just to confirm, if users want to take advantage of borrowed attributes, they can bypass the builder pattern and directly initialize the `InstrumentationScope` struct.

File: opentelemetry/src/common.rs
Comment (lalitb) [reply opentelemetry/src/common.rs:413]: > this should be a relative easy renaming for users, so I don't think its worth maintaining the alias for few releases.
This is part of opentelemetry and we are aiming to declare it RC in a week and stable soon after that.

Yes, but in current form, this will break their instrumentation build, which is not good. We can still remove the deprecated code after reaching RC.

File: opentelemetry/src/logs/logger.rs
Comment (cijothomas) [inline opentelemetry/src/logs/logger.rs:55]: we need to rename `library_logger` to something more indicative of its purpose...
provider.logger("name")
provider.logger_with_scope(scope);

open to suggestions.

File: opentelemetry/src/common.rs
Comment (cijothomas) [reply opentelemetry/src/common.rs:413]: > We can still remove the deprecated code after reaching RC.

You are right! However, this PR will break users pretty bad as it is not just renaming `InstrumentationLibrary`. I don't think its feasible to make this in a non-breaking way, and I feel its best to do this and unblock RC release of all APIs, so users know there won't be any further breaking change for instrumentations in the future.

(It might be possible to find ways to smoothen breaking change, but that effort might delay RC again and may not yield much value ultimately)

File: opentelemetry/src/common.rs
Comment (cijothomas) [reply opentelemetry/src/common.rs:431]: Tracked via https://github.com/open-telemetry/opentelemetry-rust/issues/2225

**Comment by cijothomas** [comment]
@stormshield-fabs Thanks a lot! I think the following are to be addressed before merge. We also need to write a migration guide in the PR description showing before/after, so users can easily (relatively) react to this.

Lets keep InstrumentationScope simple, and accept the perf cost now, given it is infrequent operation.
https://github.com/open-telemetry/opentelemetry-rust/pull/2220#discussion_r1809335125

Rename library-signal to something like signal_with_scope/better names.
https://github.com/open-telemetry/opentelemetry-rust/pull/2220#discussion_r1809701673

File: opentelemetry/src/logs/logger.rs
Comment (lalitb) [reply opentelemetry/src/logs/logger.rs:55]: `logger_with_scope` looks fine to me.

File: opentelemetry-stdout/examples/basic.rs
Comment (stormshield-fabs) [reply opentelemetry-stdout/examples/basic.rs]: I've reverted the `attributes` type, but this raises the question of using `Cow` for the other struct members.

File: opentelemetry/src/metrics/meter.rs
Comment (stormshield-fabs) [reply opentelemetry/src/metrics/meter.rs:28]: fixed

File: opentelemetry-sdk/src/metrics/meter_provider.rs
Comment (stormshield-fabs) [reply opentelemetry-sdk/src/metrics/meter_provider.rs:155]: fixed

File: opentelemetry/src/common.rs
Comment (stormshield-fabs) [reply opentelemetry/src/common.rs:431]: fixed

File: opentelemetry/src/common.rs
Comment (stormshield-fabs) [reply opentelemetry/src/common.rs:538]: Reverted `attributes` to its initial type as per comments above.

**Review by cijothomas — APPROVED** [review]
LGTM.
Could you modify the PR description to show the before/after example for logs/metrics/traces, so users can easily do the migration?

**Comment by cijothomas** [comment]
@stormshield-fabs oops we just cause some merge conflicts! Could you resolve them.
We can merge with one more approval.
(Also, please update pr description to show before/after code snippets to make it easy for users to migrate)

File: opentelemetry/src/common.rs
Comment (cijothomas) [reply opentelemetry/src/common.rs:413]: From SIG meeting discussion:
1. We acknowledge the amount of breaking changes
2. We are delaying the 0.27 release by a week, to ensure we can finally declare `opentelemetry` crates logs, metrics, common as RC, to assure end users that the breaking changes are ending finally!
3. It is still technically possible to make breaking changes after RC, but the bar is extremely high.
4. We also should be declaring the `opentelemetry` metrics,logs,common things stable in the next ~1 month or so.
5. Because of this, its best to accept the cost of this breaking change now.

File: opentelemetry-sdk/src/logs/log_emitter.rs
Comment (utpilla) [inline opentelemetry-sdk/src/logs/log_emitter.rs:64]: Are we changing the behavior to not override empty names with the default component name?

File: opentelemetry-sdk/src/trace/provider.rs
Comment (utpilla) [inline opentelemetry-sdk/src/trace/provider.rs:25]: Same as this https://github.com/open-telemetry/opentelemetry-rust/pull/2220/files#r1811937347

File: opentelemetry/src/global/trace.rs
Comment (utpilla) [inline opentelemetry/src/global/trace.rs:411]:
```suggestion
/// let tracer = tracer_with_scope(scope);
```

File: opentelemetry-sdk/src/lib.rs
Comment (lalitb) [inline opentelemetry-sdk/src/lib.rs:150]: I believe we can remove this aliasing. Or atleast not expose this to end-user, if we want to use it internally in sdk. I understand this is coming from instrumentation.rs, but is not necessary.

File: opentelemetry-sdk/src/logs/log_emitter.rs
Comment (stormshield-fabs) [reply opentelemetry-sdk/src/logs/log_emitter.rs:64]: I suggest we move the `signal` methods to `named_signal` and let the trait implementers handle the logic around empty names in their implementations of `signal`:

```rust
trait LoggerProvider {
    fn named_logger(&self, name: impl Into<Cow<'static, str>>) -> Self::Logger {
        let scope = InstrumentationScope::builder(name).build();
        self.logger_with_scope(scope)
    }
    
    fn logger(&self, name: impl Into<Cow<'static, str>>) -> Self::Logger;

    fn logger_with_scope(&self, scope: InstrumentationScope) -> Self::Logger;
}
```

Or maybe we should simply get rid of `named_signal`?

Comment (stormshield-fabs) [reply opentelemetry-sdk/src/logs/log_emitter.rs:64]: I went with the overrides in a separate commit, let me know what you think!

File: opentelemetry-sdk/src/lib.rs
Comment (stormshield-fabs) [reply opentelemetry-sdk/src/lib.rs]: done

File: opentelemetry-sdk/src/metrics/meter_provider.rs
Comment (utpilla) [inline opentelemetry-sdk/src/metrics/meter_provider.rs:148]:
```suggestion
/// Default meter name if empty string is provided.
```

File: opentelemetry/src/global/metrics.rs
Comment (utpilla) [inline opentelemetry/src/global/metrics.rs:41]:
```suggestion
/// Creates a [`Meter`] with the given instrumentation scope.
```

File: opentelemetry/src/global/metrics.rs
Comment (utpilla) [inline opentelemetry/src/global/metrics.rs:43]: nit:

```suggestion
/// This is a simpler alternative to `global::meter_provider().meter_with_scope(...)`
```

File: opentelemetry/src/global/mod.rs
Comment (utpilla) [inline opentelemetry/src/global/mod.rs:67]:
```suggestion
//!     let tracer = global::tracer_with_scope(scope);
```

File: opentelemetry/src/global/trace.rs
Comment (utpilla) [inline opentelemetry/src/global/trace.rs:395]: nit:
```suggestion
/// This is a simpler alternative to `global::tracer_provider().tracer_with_scope(...)`
```

File: opentelemetry/src/logs/logger.rs
Comment (utpilla) [inline opentelemetry/src/logs/logger.rs:33]:
```suggestion
    /// Returns a new logger with the given instrumentation scope.
```

File: opentelemetry/src/global/trace.rs
Comment (utpilla) [inline opentelemetry/src/global/trace.rs:392]:
```suggestion
/// Creates a [`Tracer`] with the given instrumentation scope
```

File: opentelemetry/src/metrics/meter.rs
Comment (utpilla) [inline opentelemetry/src/metrics/meter.rs:39]: nit:

It would be nice to be consistent with the ordering of these methods. Both `LoggerProvider` and `TracerProvider` traits have the simple function (`logger`, `tracer`) defined declared before the `signal_with_scope` method.

**Review by utpilla — APPROVED** [review]
Left some minor suggestions.

Great work @stormshield-fabs!!

File: opentelemetry-appender-log/src/lib.rs
Comment (lalitb) [inline opentelemetry-appender-log/src/lib.rs:172]: This scope name should be `opentelemetry-appender-log` , to be consistent with [tracing appender](https://github.com/open-telemetry/opentelemetry-rust/blob/8bd529a6d629aff7482b875cfc39275a8a71eaeb/opentelemetry-appender-tracing/src/layer.rs#L13), so as to follow the convention of `opentelemetry-appender-<name>`. 
But this is not related to PR, I will raise a separate PR for this.

File: opentelemetry-sdk/src/export/logs/mod.rs
Comment (lalitb) [inline opentelemetry-sdk/src/export/logs/mod.rs:23]: nit - change line 14, 18 (and few more) in this file to indicate `instrumentation scope` instead of `instrumentation libraries`.

**Review by lalitb — APPROVED** [review]
LGTM. Thanks for the changes.

**Comment by cijothomas** [comment]
@stormshield-fabs Can you resolve conflicts and address the nits, and we can merge!

As suggested offline, if you can add a migration guide in the PR description, similar to https://github.com/open-telemetry/opentelemetry-rust/pull/2221, that'd be awesome!

(The next release is a flood of breaking changes, so we need to write an overall migration guide, which will refer to the sub-migration-guides within each PR description.)

**Comment by lalitb** [comment]
@stormshield-fabs  I took the liberty of fixing minor merge conflicts in this PR, as it was introduced by changes from my own PR. Hope that's fine :)

**Pull request merged**
